### PR TITLE
8340420: ZGC: Should call `vm_shutdown_during_initialization` if initialization fails

### PR DESCRIPTION
**Pull request by xmas92**
ZGC does not call `vm_shutdown_during_initialization` if initialization fails during the setup of the CollectedHeap, in contrast to the other GC.

I propose we add a `ZInitialize::error` which we can use during initialisation to record errors. The first error recored is also stored and used as the error message when shutting down the VM. 

Initially used malloc to allocate the error (ed9ba5dd6805291a6b1b56566c933424230d3b4a) but feels like it is just better to have static storage for the string and not have to care about malloc potentially failing to allocate.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8340420](https://bugs.openjdk.org/browse/JDK-8340420): ZGC: Should call `vm_shutdown_during_initialization` if initialization fails (**Enhancement** - P4)

### Reviewers
 * [Stefan Karlsson](https://openjdk.org/census#stefank) (@stefank - **Reviewer**)
 * [Hamlin Li](https://openjdk.org/census#mli) (@Hamlin-Li - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/21254/head:pull/21254` \
`$ git checkout pull/21254`

Update a local copy of the PR: \
`$ git checkout pull/21254` \
`$ git pull https://git.openjdk.org/jdk.git pull/21254/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 21254`

View PR using the GUI difftool: \
`$ git pr show -t 21254`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/21254.diff">https://git.openjdk.org/jdk/pull/21254.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/21254#issuecomment-2382235880)

### CODE REVIEW

File: src/hotspot/share/gc/z/zInitialize.cpp [hunk 1]
```diff
@@ -62,3 +73,51 @@ ZInitialize::ZInitialize(ZBarrierSet* barrier_set) {
...
+  va_start(argp, msg_format);
+  const FormatBuffer<ErrorMessageLength> error(FormatBufferDummy(), msg_format, argp);
+  va_end(argp);
+  register_error(false /* debug */, error);
+}
+
+void ZInitialize::error_d(const char* msg_format, ...) {
+  va_list argp;
+  va_start(argp, msg_format);
+  const FormatBuffer<ErrorMessageLength> error(FormatBufferDummy(), msg_format, argp);
```
Comment (stefank) [inline src/hotspot/share/gc/z/zInitialize.cpp:103]: Maybe consider a name that doesn't clash with the function name.

File: src/hotspot/share/gc/z/zInitialize.cpp [hunk 2]
```diff
@@ -62,3 +73,51 @@ ZInitialize::ZInitialize(ZBarrierSet* barrier_set) {
...
+    log_error_pd(gc)("%s", error);
+  } else {
+    log_error_p(gc)("%s", error);
+  }
+}
+
+void ZInitialize::error(const char* msg_format, ...) {
+  va_list argp;
+  va_start(argp, msg_format);
+  const FormatBuffer<ErrorMessageLength> error(FormatBufferDummy(), msg_format, argp);
```
Comment (stefank) [inline src/hotspot/share/gc/z/zInitialize.cpp:95]: Maybe consider a name that doesn't clash with the function name.

File: src/hotspot/share/gc/z/zInitialize.cpp [hunk 3]
```diff
@@ -62,3 +73,51 @@ ZInitialize::ZInitialize(ZBarrierSet* barrier_set) {
...
+bool ZInitialize::had_error() {
+  return _had_error;
+}
+
+const char* ZInitialize::error_message() {
+  assert(had_error(), "Should have registered an error");
+  if (!had_error()) {
+    return "Unknown error, check error GC logs";
+  }
+  return _error_message;
```
Comment (stefank) [inline src/hotspot/share/gc/z/zInitialize.cpp:117]: Maybe invert the conditional?
```suggestion
  if (had_error()) {
    return _error_message;
  }
  return "Unknown error, check error GC logs";
```

File: src/hotspot/share/gc/z/zInitialize.hpp [hunk 4]
```diff
@@ -24,16 +24,38 @@
...
+#include "memory/allStatic.hpp"
+#include "utilities/compilerWarnings.hpp"
+
+#include <cstddef>
 
 class ZBarrierSet;
 
-class ZInitialize {
+class ZInitializer {
+ public:
```
Comment (stefank) [inline src/hotspot/share/gc/z/zInitialize.hpp:35]: Indentation doesn't follow the ZGC style.

File: src/hotspot/share/gc/z/zInitialize.hpp [hunk 5]
```diff
@@ -24,16 +24,38 @@
...
+  ZInitializer(ZBarrierSet* barrier_set);
+};
+
+class ZInitialize : public AllStatic {
 private:
-  void pd_initialize();
+  static constexpr size_t ErrorMessageLength = 256;
+  static char _error_message[ErrorMessageLength];
```
Comment (stefank) [inline src/hotspot/share/gc/z/zInitialize.hpp:42]: I would suggest that we separate the constants
```suggestion
  static constexpr size_t ErrorMessageLength = 256;

  static char _error_message[ErrorMessageLength];
```

**Review by stefank — CHANGES_REQUESTED** [review]

**Review by stefank — APPROVED** [review]

**Review by Hamlin-Li — APPROVED** [review]

**Comment by xmas92** [comment]
Thanks for the reviews.
/integrate

**Pull request closed without merging**